### ISSUE TITLE
v2.0.0 is extremely slow in comparaison of v1.1.1

### ISSUE THREAD

**Issue by zoontek** [simonhaenisch/prettier-plugin-organize-imports#19]
Hi 👋

First, thanks for this amazing project, we use it on all our codebases!
I noticed that since the latest update, the formatting (with `prettier --write`) is extremely slow to perform.

**Before (1.1.1):**

<img width="1792" alt="Screenshot 2021-05-12 at 15 27 43" src="https://user-images.githubusercontent.com/1902323/117983372-2e856780-b337-11eb-94c1-009458948f9e.png">

**After (2.0.0):**

<img width="1792" alt="Screenshot 2021-05-12 at 15 27 31" src="https://user-images.githubusercontent.com/1902323/117983432-3ba25680-b337-11eb-8888-1d24bab2a93d.png">

**Comment by simonhaenisch** [comment]
Hi, thanks for the info. Just so you know, as long as you're only using TypeScript, there's no harm staying on 1.1.1 for now. There's only been two changes really, one was to add support for two more parsers (`babel` and `babel-ts`), the other was to change some methods used from the TypeScript API (9d88c9b4ed03e4e41dc13774d100a3584d75e3ec).

I'll investigate which change this came from asap!

**Comment by simonhaenisch** [comment]
So i've done some investigation by adding timers to the test wrapping the `prettier.format()` call, and weirdly I get these timings with `1.1.1`:

```
prettify: 432.327ms
prettify: 9.994ms
prettify: 4.987ms
prettify: 14.623ms
prettify: 7.875ms
prettify: 5.295ms
```

and with `2.0.0`:

```
prettify: 500.186ms
prettify: 10.554ms
prettify: 5.854ms
prettify: 16.735ms
prettify: 8.895ms
prettify: 5.589ms
```

Even though there's a performance impact, it's not that significant. However what's huge is always the first call, and it seems like a lot of the startup time stems from loading plugins (see https://github.com/prettier/prettier/issues/4776#issuecomment-401523480). When I remove the plugin from the config, the first time gets cut in half (`226 ms`). So I checked my plugin code and the majority of the time is taken by importing `prettier/parser-typescript` and `typescript`, both of which were already imported in `v1.1.1`.

So now my question is whether you changed anything else other than updating the plugin? It seems to me like what's different is that Prettier restarts for every file it formats, rather than starting up once and then re-using the instance. I'm not sure how any of the changes between 1.1.1 and 2.0 would cause this 🤷🏻‍♂️

**Comment by simonhaenisch** [comment]
I just tried on one of my own projects, and with 1.1.1 I get this:

```
❯ npx prettier src/**/*.ts --write
src/cli.ts 269ms
src/index.ts 65ms
src/lib/config.ts 62ms
src/lib/generate-output.ts 72ms
src/lib/get-html.ts 15ms
src/lib/get-marked-with-highlighter.ts 32ms
src/lib/get-output-file-path.ts 11ms
src/lib/help.ts 39ms
src/lib/helpers.ts 29ms
src/lib/is-http-url.ts 8ms
src/lib/is-md-file.ts 5ms
src/lib/md-to-pdf.ts 56ms
src/lib/read-file.ts 15ms
src/lib/serve-dir.ts 23ms
src/lib/validate-node-version.ts 10ms
src/test/api.spec.ts 47ms
src/test/api.type-spec.ts 14ms
src/test/cli.spec.ts 57ms
src/test/lib.spec.ts 81ms
```

and just after only upgrading the plugin to `2.0.0`:

```
❯ npx prettier src/**/*.ts --write
src/cli.ts 284ms
src/index.ts 78ms
src/lib/config.ts 82ms
src/lib/generate-output.ts 63ms
src/lib/get-html.ts 16ms
src/lib/get-marked-with-highlighter.ts 22ms
src/lib/get-output-file-path.ts 14ms
src/lib/help.ts 46ms
src/lib/helpers.ts 35ms
src/lib/is-http-url.ts 13ms
src/lib/is-md-file.ts 9ms
src/lib/md-to-pdf.ts 64ms
src/lib/read-file.ts 18ms
src/lib/serve-dir.ts 28ms
src/lib/validate-node-version.ts 12ms
src/test/api.spec.ts 48ms
src/test/api.type-spec.ts 21ms
src/test/cli.spec.ts 56ms
src/test/lib.spec.ts 93ms
```

So there's really not that much of a difference just from the plugin. Can you maybe try to `npm i prettier@latest` as well? Are there any other plugins you are using?

**Comment by zoontek** [comment]
I use prettier `2.2.1` + prettier-plugin-organize-imports `1.1.1` / `2.0.0`.
I tried using prettier `2.3.0` (latest), without success. I will investigate on my side. 🤔

**Comment by blidblid** [comment]
Love the plugin, but it slows down Prettier 10x for me. Running Prettier on a repo went from 2.5 seconds to 26 seconds with the plugin enabled.

So I had a look at the node process when this plugin in runs, it spends obsene amounts of time [here](https://github.com/simonhaenisch/prettier-plugin-organize-imports/blob/master/index.js#L64). 

![image](https://user-images.githubusercontent.com/30099577/123259532-daa09d80-d4f4-11eb-9c7b-58979e3bcbfa.png)

That helper seems to do a lot of work

> A root directory to resolve relative path entries in the config file to. e.g. outDir

meaning it traverses the whole repository and resolves `compilerOptions.paths`. Refactoring that particular line would speed up this plugin by a ton.

**Comment by simonhaenisch** [comment]
Hey @blidblid thanks a lot for looking into this!

I actually noticed that there's some potential for improvements because the plugin is creating a new language service every time it's run, i. e. for each file. The problem I have is that I don't know much about TypeScript language service hosts and I've been struggling to find documentation or good examples.

So it's quite helpful that you've narrowed down the bottleneck... it seems like there's a relatively simple work-around which is to have a `Map` of the tsconfigs that have already been read, and then not call `parseJsonConfigFileContent` again but use the cache instead. I'll see that I can get this implemented.

**Comment by simonhaenisch** [comment]
Just published `2.2.0` can you give it a try please?

**Comment by blidblid** [comment]
`2.2.0` runs the through repo in 6 seconds, compared to 26 in seconds `2.1.0`.  Thanks a lot for the quick fix!
